lib/ansible/executor/task_executor.py

```diff
@@ -423,41 +431,55 @@ def _execute(self, variables=None):
 
         context_validation_error = None
         try:
-            # TODO: remove play_context as this does not take delegation into account, task itself should hold values
-            #  for connection/shell/become/terminal plugin options to finalize.
-            #  Kept for now for backwards compatibility and a few functions that are still exclusive to it.
-
-            # apply the given task's information to the connection info,
-            # which may override some fields already set by the play or
-            # the options specified on the command line
-            self._play_context = self._play_context.set_task_and_variable_override(task=self._task, variables=variables, templar=templar)
-
-            # fields set from the play/task may be based on variables, so we have to
-            # do the same kind of post validation step on it here before we use it.
-            self._play_context.post_validate(templar=templar)
-
-            # now that the play context is finalized, if the remote_addr is not set
-            # default to using the host's address field as the remote address
-            if not self._play_context.remote_addr:
-                self._play_context.remote_addr = self._host.address
-
-            # We also add "magic" variables back into the variables dict to make sure
-            # a certain subset of variables exist.
-            self._play_context.update_vars(variables)
-
-        except AnsibleError as e:
-            # save the error, which we'll raise later if we don't end up
-            # skipping this task during the conditional evaluation step
-            context_validation_error = e
-
-        # Evaluate the conditional (if any) for this task, which we do before running
-        # the final task post-validation. We do this before the post validation due to
-        # the fact that the conditional may specify that the task be skipped due to a
-        # variable not being present which would otherwise cause validation to fail
-        try:
-            if not self._task.evaluate_conditional(templar, variables):
-                display.debug("when evaluation is False, skipping this task")
-                return dict(changed=False, skipped=True, skip_reason='Conditional result was False', _ansible_no_log=self._play_context.no_log)
+            if from_loop:
```

Comment:
My initial idea was to extract each step that we are going to move out of `_execute` and put them into own methods and use them where needed, it might be clearer that special casing `from_loop`.